### PR TITLE
add sdio driver for GD32F4

### PR DESCRIPTION
**Pull request by GD32-MCU**
## Summary
add sdio driver for gd32f4

### CODE REVIEW

File: arch/arm/src/gd32f4/gd32f4xx_sdio.c [hunk 1]
```diff
@@ -0,0 +1,3070 @@
...
+#  endif
+#  if defined(CONFIG_GD32F4_GD32F4XX)
+#    if (CONFIG_GD32_SDIO_DMAPRIO & ~DMA_CHXCTL_PRIO_MASK) != 0
+#      error "Illegal value for CONFIG_GD32_SDIO_DMAPRIO"
+#    endif
+#  else
+#    error "Unknown GD32 DMA"
+#  endif
+#else
+#  undef CONFIG_GD32_SDIO_DMAPRIO
```
Comment (acassis) [inline arch/arm/src/gd32f4/gd32f4xx_sdio.c:104]: Please remove the code below "#error ", it will not be executed anyway.

Also, please change the error message to:
```
#  error "SDIO DMA Support not implemented yet!"
```

File: arch/arm/src/gd32f4/gd32f4xx_sdio.c [hunk 2]
```diff
@@ -0,0 +1,3070 @@
...
+{
+  regs->power   = (uint8_t)getreg32(GD32_SDIO_PWRCTL);
+  regs->clkctl   = (uint16_t)getreg32(GD32_SDIO_CLKCTL);
+  regs->datactl   = (uint16_t)getreg32(GD32_SDIO_DATACTL);
+  regs->datato  = getreg32(GD32_SDIO_DATATO);
+  regs->datalen    = getreg32(GD32_SDIO_DATALEN);
+  regs->datacnt  = getreg32(GD32_SDIO_DATACNT);
+  regs->stat     = getreg32(GD32_SDIO_STAT);
+  regs->inten    = getreg32(GD32_SDIO_INTEN);
+  regs->fifocnt = getreg32(GD32_SDIO_FIFOCNT);
```
Comment (acassis) [inline arch/arm/src/gd32f4/gd32f4xx_sdio.c:726]: Please align to all "=" be at same position, see stm32_sdiosample() for reference

File: arch/arm/src/gd32f4/gd32f4xx_sdio.c [hunk 3]
```diff
@@ -0,0 +1,3070 @@
...
+         DEBUG_NSAMPLES * sizeof(struct gd32_sampleregs_s));
+}
+#endif
+
+/****************************************************************************
+ * Name: gd32_sdiosample
+ *
+ * Description:
+ *   Sample SDIO registers
+ *
```
Comment (acassis) [inline arch/arm/src/gd32f4/gd32f4xx_sdio.c:712]: Please include Input Parameters and Returned Value.

File: arch/arm/src/gd32f4/gd32f4xx_sdio.c [hunk 4]
```diff
@@ -0,0 +1,3070 @@
...
+  regs->fifocnt = getreg32(GD32_SDIO_FIFOCNT);
+}
+#endif
+
+/****************************************************************************
+ * Name: gd32_sample
+ *
+ * Description:
+ *   Sample SDIO/DMA registers
+ *
```
Comment (acassis) [inline arch/arm/src/gd32f4/gd32f4xx_sdio.c:735]: Please include Input Parameters and Returned Values.

File: arch/arm/src/gd32f4/gd32f4xx_sdio.c [hunk 5]
```diff
@@ -0,0 +1,3070 @@
...
+  gd32_sdiosample(&regs->sdio);
+}
+#endif
+
+/****************************************************************************
+ * Name: gd32_sdiodump
+ *
+ * Description:
+ *   Dump one register sample
+ *
```
Comment (acassis) [inline arch/arm/src/gd32f4/gd32f4xx_sdio.c:759]: Please include Input Parameters and Returned Values:

File: arch/arm/src/gd32f4/gd32f4xx_sdio.c [hunk 6]
```diff
@@ -0,0 +1,3070 @@
...
+  mcinfo("FIFOCNT[%08x]: %08x\n", GD32_SDIO_FIFOCNT, regs->fifocnt);
+}
+#endif
+
+/****************************************************************************
+ * Name: gd32_dumpsample
+ *
+ * Description:
+ *   Dump one register sample
+ *
```
Comment (acassis) [inline arch/arm/src/gd32f4/gd32f4xx_sdio.c:783]: Please include Input Parameters and Returned Values:

File: arch/arm/src/gd32f4/gd32f4xx_sdio.c [hunk 7]
```diff
@@ -0,0 +1,3070 @@
...
+  gd32_sdiodump(&regs->sdio, msg);
+}
+#endif
+
+/****************************************************************************
+ * Name: gd32_dumpsamples
+ *
+ * Description:
+ *   Dump all sampled register data
+ *
```
Comment (acassis) [inline arch/arm/src/gd32f4/gd32f4xx_sdio.c:807]: Add Input Parameters and Returned Value.

File: arch/arm/src/gd32f4/gd32f4xx_sdio.c [hunk 8]
```diff
@@ -0,0 +1,3070 @@
...
+#endif
+}
+#endif
+
+/****************************************************************************
+ * Name: gd32_dmacallback
+ *
+ * Description:
+ *   Called when SDIO DMA completes
+ *
```
Comment (acassis) [inline arch/arm/src/gd32f4/gd32f4xx_sdio.c:844]: Add Input Parameters and Returned Values.

File: arch/arm/src/gd32f4/gd32f4xx_sdio.c [hunk 9]
```diff
@@ -0,0 +1,3070 @@
...
+}
+#endif
+
+/****************************************************************************
+ * Name: gd32_log2
+ *
+ * Description:
+ *   Take (approximate) log base 2 of the provided number (Only works if the
+ *   provided number is a power of 2).
+ *
```
Comment (acassis) [inline arch/arm/src/gd32f4/gd32f4xx_sdio.c:895]: Add Input Parameter and Returned Value.

File: arch/arm/src/gd32f4/gd32f4xx_sdio.c [hunk 10]
```diff
@@ -0,0 +1,3070 @@
...
+
+  return log2;
+}
+
+/****************************************************************************
+ * Name: gd32_data_config
+ *
+ * Description:
+ *   Configure the SDIO data path for the next data transfer
+ *
```
Comment (acassis) [inline arch/arm/src/gd32f4/gd32f4xx_sdio.c:925]: Add Input Parameters and Returned Values:

File: arch/arm/src/gd32f4/gd32f4xx_sdio.c [hunk 11]
```diff
@@ -0,0 +1,3070 @@
...
+}
+
+/****************************************************************************
+ * Name: gd32_interrupt
+ *
+ * Description:
+ *   SDIO interrupt handler
+ *
+ * Input Parameters:
+ *   dev - An instance of the SDIO device interface
```
Comment (acassis) [inline arch/arm/src/gd32f4/gd32f4xx_sdio.c:1264]: Input parameters is incorrect

File: arch/arm/src/gd32f4/gd32f4xx_sdio.c [hunk 12]
```diff
@@ -0,0 +1,3070 @@
...
+ *
+ * Description:
+ *   Receive response to SDIO command.  Only the critical payload is
+ *   returned -- that is 32 bits for 48 bit status and 128 bits for 136 bit
+ *   status.  The driver implementation should verify the correctness of
+ *   the remaining, non-returned bits (CRCs, CMD index, etc.).
+ *
+ * Input Parameters:
+ *   dev    - An instance of the SDIO device interface
+ *   Rx - Buffer in which to receive the response
```
Comment (acassis) [inline arch/arm/src/gd32f4/gd32f4xx_sdio.c:2096]: Input Parameters is incorrect, please fix

File: arch/arm/src/gd32f4/gd32f4xx_sdio.c [hunk 13]
```diff
@@ -0,0 +1,3070 @@
...
+ *   setup and the subsequent events.
+ *
+ *   The enabled events persist until either (1) SDIO_WAITENABLE is called
+ *   again specifying a different set of wait events, or (2) SDIO_EVENTWAIT
+ *   returns.
+ *
+ * Input Parameters:
+ *   dev      - An instance of the SDIO device interface
+ *   eventset - A bitset of events to enable or disable (see SDIOWAIT_*
+ *              definitions). 0=disable; 1=enable.
```
Comment (acassis) [inline arch/arm/src/gd32f4/gd32f4xx_sdio.c:2318]: Input Parameters missing timeout

File: arch/arm/src/gd32f4/gd32f4xx_sdio.c [hunk 14]
```diff
@@ -0,0 +1,3070 @@
...
+ *
+ * Description:
+ *   Wait for one of the enabled events to occur (or a timeout).  Note that
+ *   all events enabled by SDIO_WAITEVENTS are disabled when gd32_event_wait
+ *   returns.  SDIO_WAITEVENTS must be called again before gd32_event_wait
+ *   can be used again.
+ *
+ * Input Parameters:
+ *   dev     - An instance of the SDIO device interface
+ *   timeout - Maximum time in milliseconds to wait.  Zero means immediate
```
Comment (acassis) [inline arch/arm/src/gd32f4/gd32f4xx_sdio.c:2405]: Remove timeout

File: arch/arm/src/gd32f4/gd32f4xx_sdio.c [hunk 15]
```diff
@@ -0,0 +1,3070 @@
...
+ * Name: gd32_callback
+ *
+ * Description:
+ *   Perform callback.
+ *
+ * Assumptions:
+ *   This function does not execute in the context of an interrupt handler.
+ *   It may be invoked on any user thread or scheduled on the work thread
+ *   from an interrupt handler.
+ *
```
Comment (acassis) [inline arch/arm/src/gd32f4/gd32f4xx_sdio.c:2799]: Please include Input Parameter and Returned Value.

File: arch/arm/src/gd32f4/gd32f4xx_sdio.c [hunk 16]
```diff
@@ -0,0 +1,3070 @@
...
+
+/****************************************************************************
+ * Name: sdio_set_sdio_card_isr
+ *
+ * Description:
+ *   SDIO card generates interrupt via SDIO_DATA_1 pin.
+ *   Called by board-specific logic to register an ISR for SDIO card.
+ *
+ * Input Parameters:
+ *   func      - callback function.
```
Comment (acassis) [inline arch/arm/src/gd32f4/gd32f4xx_sdio.c:3039]: Missing dev

**Comment by acassis** [comment]
Nice work @GD32-MCU!!! Just some minor issues found (coding styles and doc functions).

File: arch/arm/src/gd32f4/Kconfig [hunk 17]
```diff
@@ -2089,7 +2093,7 @@ menu "RTC Configuration"
 config GD32F4_RTC_MAGIC_REG
 	int "BKP register"
 	default 0
-	range 0 19
+	range  0 19
```
Comment (xiaoxiang781216) [inline arch/arm/src/gd32f4/Kconfig:2096]: why change

File: arch/arm/src/gd32f4/Kconfig [hunk 18]
```diff
@@ -663,7 +667,7 @@ config ARCH_BOARD_GD32F4_CUSTOM_CLOCKCONFIG
 
 config GD32F4_TCMEXCLUDE
 	bool "Exclude TCM SRAM from the heap"
-	default y if ARCH_DMA || LIBC_ARCH_ELF
+	default y if  ARCH_DMA || LIBC_ARCH_ELF
```
Comment (xiaoxiang781216) [inline arch/arm/src/gd32f4/Kconfig:670]: why change

File: arch/arm/src/gd32f4/Kconfig [hunk 19]
```diff
@@ -2135,6 +2139,37 @@ endmenu # RTC configuration
...
+	default n
+	---help---
+		If you are using an external SDCard module that does not have the
+		pull-up resistors for the SDIO interface (like the Gadgeteer SD Card
+		Module) then enable this option to activate the internal pull-up
+		resistors.
+
+config GD32F4_SDIO_DMA
+	bool "Support DMA data transfers"
+	default y if GD32F4_DMA1
```
Comment (xiaoxiang781216) [inline arch/arm/src/gd32f4/Kconfig:2161]:
```suggestion
	default GD32F4_DMA1
```

File: arch/arm/src/gd32f4/Kconfig [hunk 20]
```diff
@@ -2135,6 +2139,37 @@ endmenu # RTC configuration
...
+	default n if !GD32F4_DMA1
+	select SDIO_DMA
+	depends on GD32F4_DMA1
```
Comment (xiaoxiang781216) [inline arch/arm/src/gd32f4/Kconfig:2164]: why both default and depend to the same GD32F4_DMA1?

File: boards/arm/gd32f4/gd32f450zk-eval/include/board.h [hunk 21]
```diff
@@ -28,7 +28,7 @@
 #include <nuttx/config.h>
 
 #ifndef __ASSEMBLY__
-#  include <stdint.h>
+# include <stdint.h>
```
Comment (xiaoxiang781216) [inline boards/arm/gd32f4/gd32f450zk-eval/include/board.h:31]: revert the change

File: boards/arm/gd32f4/gd32f450zk-eval/src/Makefile [hunk 22]
```diff
@@ -49,6 +49,10 @@ ifeq ($(CONFIG_SPI),y)
 CSRCS +=  gd32f4xx_spi.c
 endif
 
+ifeq ($(CONFIG_MMCSD),y)
+CSRCS +=  gd32f4xx_sdio.c
```
Comment (xiaoxiang781216) [inline boards/arm/gd32f4/gd32f450zk-eval/src/Makefile:53]: remove extra space after += from line 44 to 53

File: boards/arm/gd32f4/gd32f450zk-eval/include/board.h [hunk 23]
```diff
@@ -358,4 +358,26 @@ typedef enum
...
+
+#define GPIO_SDIO_CMD_PIN   GPIO_SDIO_CMD_2
+#define GPIO_SDIO_CLK_PIN   GPIO_SDIO_CK_2
+#define GPIO_SDIO_DAT0_PIN  GPIO_SDIO_D0_2 
+#define GPIO_SDIO_DAT1_PIN  GPIO_SDIO_D1_3
+#define GPIO_SDIO_DAT2_PIN  GPIO_SDIO_D2_3
+#define GPIO_SDIO_DAT3_PIN  GPIO_SDIO_D3
+
+#ifdef CONFIG_GD32F4_SDIO_DMA
+#  define SDIO_DMA_INTEN            (DMA_CHXCTL_SDEIE | DMA_CHXCTL_TAEIE | DMA_CHXCTL_FTFIE)
```
Comment (xiaoxiang781216) [inline boards/arm/gd32f4/gd32f450zk-eval/include/board.h:380]:
```suggestion
#  define SDIO_DMA_INTEN   (DMA_CHXCTL_SDEIE | DMA_CHXCTL_TAEIE | DMA_CHXCTL_FTFIE)
```

File: boards/arm/gd32f4/gd32f450zk-eval/include/board.h [hunk 24]
```diff
@@ -28,7 +28,7 @@
 #include <nuttx/config.h>
 
 #ifndef __ASSEMBLY__
-#  include <stdint.h>
+#include <stdint.h>
```
Comment (xiaoxiang781216) [inline boards/arm/gd32f4/gd32f450zk-eval/include/board.h:31]: revert the change

Comment (xiaoxiang781216) [reply boards/arm/gd32f4/gd32f450zk-eval/include/board.h:31]: @GD32-MCU but this chagne doesn't revert yet.

Comment (GD32-MCU) [reply boards/arm/gd32f4/gd32f450zk-eval/include/board.h]: Thanks for your  review,  now the chagne is revert

File: arch/arm/src/gd32f4/gd32f4xx_sdio.c [hunk 25]
```diff
@@ -0,0 +1,3133 @@
...
+ * Private Types
+ ****************************************************************************/
+
+  /* rcu  peripheral clock */
+
+typedef enum
+{
+    ENABLE,
+    DISABLE,
+}periph_enum;
```
Comment (pkarashchenko) [inline arch/arm/src/gd32f4/gd32f4xx_sdio.c:232]:
```suggestion
typedef enum
{
  ENABLE,
  DISABLE,
} periph_enum;
```

File: arch/arm/src/gd32f4/gd32f4xx_sdio.c [hunk 26]
```diff
@@ -0,0 +1,3133 @@
...
+#define SDIO_WAITALL_INTC   (SDIO_CMDDONE_INTC | SDIO_RESPDONE_INTC | \
+                            SDIO_XFRDONE_INTC | SDIO_INTC_DTBLKENDC)
+
+/* Let's wait until we have both SDIO transfer complete and DMA complete. */
+
+#define SDIO_XFRDONE_FLAG  (1)
+#define SDIO_DMADONE_FLAG  (2)
+#define SDIO_ALLDONE       (3)
+
+#define GD32_SDIO_USE_DEFAULT_BLOCKSIZE ((uint8_t)-1)
```
Comment (pkarashchenko) [inline arch/arm/src/gd32f4/gd32f4xx_sdio.c:220]: Why not `UINT8_MAX` or `255u`?

File: arch/arm/src/gd32f4/gd32f4xx_sdio.c [hunk 27]
```diff
@@ -0,0 +1,3133 @@
...
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_SDIO_XFRDEBUG
+static void gd32_dumpsample(struct gd32_dev_s *priv,
+                             struct gd32_sampleregs_s *regs,
+                             const char *msg)
```
Comment (pkarashchenko) [inline arch/arm/src/gd32f4/gd32f4xx_sdio.c:814]:
```suggestion
static void gd32_dumpsample(struct gd32_dev_s *priv,
                            struct gd32_sampleregs_s *regs,
                            const char *msg)
```

File: arch/arm/src/gd32f4/gd32f4xx_sdio.c [hunk 28]
```diff
@@ -0,0 +1,3133 @@
...
+{
+#if defined(CONFIG_DEBUG_DMA_INFO) && defined(CONFIG_GD32F4_SDIO_DMA)
+  if (priv->dmamode)
+    {
+      gd32_dma_sample (priv->dma, &regs->dma, msg);
```
Comment (pkarashchenko) [inline arch/arm/src/gd32f4/gd32f4xx_sdio.c:819]:
```suggestion
      gd32_dma_sample(priv->dma, &regs->dma, msg);
```

File: arch/arm/src/gd32f4/gd32f4xx_sdio.c [hunk 29]
```diff
@@ -0,0 +1,3133 @@
...
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_SDIO_XFRDEBUG
+static void gd32_dumpsamples(struct gd32_dev_s *priv)
+{
+  gd32_dumpsample(priv, &g_sampleregs[SAMPLENDX_BEFORE_SETUP],
+                   "Before setup");
```
Comment (pkarashchenko) [inline arch/arm/src/gd32f4/gd32f4xx_sdio.c:845]:
```suggestion
  gd32_dumpsample(priv, &g_sampleregs[SAMPLENDX_BEFORE_SETUP],
                  "Before setup");
```
same in other similar lines below

File: arch/arm/src/gd32f4/gd32f4xx_sdio.c [hunk 30]
```diff
@@ -0,0 +1,3133 @@
...
+    {
+      sdio_clk = sdio_clk / (2 + clkdiv);
+    }
+
+  /*  Convert Timeout in Ms to SD_CLK counts */
+
+  timeout  = timeout * (sdio_clk / 1000);
+
+  putreg32(timeout,    GD32_SDIO_DATATO);  /* Set DTIMER */
+  putreg32(dlen,       GD32_SDIO_DATALEN); /* Set DATALEN */
```
[the remark this answers is not in the document]
Comment (GD32-MCU) [reply arch/arm/src/gd32f4/gd32f4xx_sdio.c]: Thanks for the review

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by xiaoxiang781216 — APPROVED** [review]

File: boards/arm/gd32f4/gd32f450zk-eval/src/gd32f4xx_sdio.c [hunk 31]
```diff
@@ -0,0 +1,156 @@
...
+#  undef HAVE_NCD
+#endif
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static struct sdio_dev_s *g_sdio_dev;
+#ifdef HAVE_NCD
+static bool g_sd_inserted = 0xff; /* Impossible value */
```
Comment (pkarashchenko) [inline boards/arm/gd32f4/gd32f450zk-eval/src/gd32f4xx_sdio.c:60]: I think type here should be changed to `uint8_t` as I think C99 will transform such construction to `true`
